### ISSUE TITLE
Question: Examples vulkan mislead 

### ISSUE THREAD

**Issue by amerkoleci** [KhronosGroup/Vulkan-Samples#365]
Hi,
I saw that example framework is using hash and creates shaders, pipeline layout, pipelines etc on demand during frame rendering/dispatch, isn't this 'bad' practice and it's better to create them on loading?

Just curious if I'm not seeing something about correct vulkan usage.

Feel free to close issue if it doesn't reflect project ideas.

Thanks

**Comment by gpx1000** [comment]
Hi,

Thanks for the question.  However, I'm not certain I fully know what you're talking about.  As far as I know, the samples framework does create shaders, pipelines, and pipeline layout only once during initialization or resize.
While the framework could use some tweaking to show more best practices in other areas, and could get some improvements.  On the whole it's pretty good.

Thanks

**Comment by amerkoleci** [comment]
What I'm talking is: https://github.com/KhronosGroup/Vulkan-Samples/blob/master/framework/rendering/pipeline_state.cpp, the commans buffer makes heavy use and set the shaders, states (rasterization/blend/depth) and before drawing/dispatch it flush and creates pipeline on demand.

The programming model seams similar to d3d11 for example.

**Comment by gpx1000** [comment]
I'm not the right person to explain the framework.  Especially as I'm not certain I'm seeing exactly what you're talking about.  Thus I'll let others that know the framework better than I a chance to review this question.

**Comment by TomAtkinsonArm** [comment]
Hi @amerkoleci,

I completely agree with your points... There are some aspects of the framework which may be over engineered and may need revisiting. The main blocking factor from revisiting this is that the rendering part of the framework is the core part of this project and so any large changes would need to propagate through the entire project

Its worth noting before I start, I will need to revisit this part of the code base in more depth to verify that my claims are all true, this is my opinion since the last time I reviewed this part of the framework.

Our `vkb::CommandBuffer` performs a find and create of resources from the pipeline cache before recording to the command buffer. This means at the first frame, all resources are created once and then later frames are taken from the pipeline cache. The pipeline cache is then cleaned when there is a resize event or other breaking event which the application needs to react too.

This pattern isn't inherently bad, there is a hybrid build and draw phase and the rest of the frames simply draw. One issue I see with this is the granularity of the pipeline cache. We need to perform lots of look ups which by themselves have low overhead but in the thousands will effect the time taken to render a frame.

One possible solution, which may be easier than an entire rewrite of the rendering framework, would be to group common resources into structs. For instance have a `rendering::Mesh` struct which contains all the handles and Vulkan resources needed to render a mesh. This would cut down the amount of pipeline cache look ups.

In short, I think your concern is justified and there are better ways of doing this. The projects goals are to show the differences between bad and good practices for developers and so this type of refactor is lower on the priority list than covering some of the new API features and graphics techniques which are soon to be released

Let me know if this helps

Tom

**Comment by SaschaWillems** [comment]
Closing this due to inactivity.
